### ISSUE TITLE
Identifying scikit-image incompatibilities with dask input arrays

### ISSUE THREAD

**Issue by GenevieveBuckley** [scikit-image/scikit-image#3846]
## Description

We don't currently have a good sense of where the scikit-image library is incompatible with dask, so I've adapted @emmanuelle's [script](https://gist.github.com/emmanuelle/1964c4a81fabebd41173d75ff1e26018) for checking 3D compatibility in scikit-image for this purpose. I'd like to hear opinions from the community on where efforts to improve compatibility should be focused.

When we talk about dask & scikit-image there are basically three levels:

1. **Incompatible**: An error is raised, nothing works. This is what we want to minimise for dask arrays in scikit-image.
2. **Compatible**: No error is raised. You can do it, but it may or may not be a sensible idea. Computation happens immediately, and typically a numpy result is returned (which might cause memory issues, depending on the size of the input & output data). This is most likely what scikit-image should be aiming for.
3. **Designed for dask**: All the advantages of dask including lazy computation, etc. Virtually none of scikit-image falls into this category.

Here is a list* of scikit-image functions that work with numpy arrays, but do not work with a dask input array. *This list is incomplete, may contain inaccuracies, and ignores any function that takes multiple input arguments. Additions and corrections welcome!

**skimage.exposure**
- [ ] skimage.exposure.equalize_hist

**skimage.feature**
- [ ] skimage.feature.canny
- [ ] skimage.feature.corner_peaks
- [ ] skimage.feature.peak_local_max

**skimage.filters**
- [ ] skimage.filters.rank_order
- [ ] skimage.filters.threshold_li

**skimage.measure**
- [ ] skimage.measure.marching_cubes_lewiner
- [ ] skimage.measure.moments_hu

**skimage.morphology**
- [ ] skimage.morphology.area_closing
- [ ] skimage.morphology.area_opening
- [ ] skimage.morphology.black_tophat
- [ ] skimage.morphology.diameter_closing
- [ ] skimage.morphology.diameter_opening
- [ ] skimage.morphology.max_tree
- [ ] skimage.morphology.max_tree_local_maxima
- [ ] skimage.morphology.medial_axis
- [ ] skimage.morphology.remove_small_holes
- [ ] skimage.morphology.remove_small_objects
- [ ] skimage.morphology.skeletonize
- [ ] skimage.morphology.white_tophat

**skimage.segmentation**
- [ ] skimage.segmentation.chan_vese
- [ ] skimage.segmentation.clear_border

**skimage.transform**
- [ ] skimage.transform.frt2
- [ ] skimage.transform.hough_ellipse
- [ ] skimage.transform.hough_line
- [ ] skimage.transform.ifrt2
- [ ] skimage.transform.iradon_sart
- [ ] skimage.transform.probabilistic_hough_line
- [ ] skimage.transform.radon

## Version information
```python
# Paste the output of the following python commands
from __future__ import print_function
import sys; print(sys.version)
import platform; print(platform.platform())
import skimage; print("scikit-image version: {}".format(skimage.__version__))
import numpy; print("numpy version: {}".format(numpy.__version__))
```

```python
3.7.3 (default, Mar 27 2019, 17:13:21) [MSC v.1915 64 bit (AMD64)]
Windows-10-10.0.17763-SP0
scikit-image version: 0.16.dev0
numpy version: 1.16.2
dask version: 1.2.0
```

**Comment by hmaarrfk** [comment]
Are we trying to anticipate the need for users to use dask?

Have we had anybody report this kind of incompatibility or is this a curiosity from the side of the core devs?

I'm affraid that without a particular application in mind, we might end up slowing things down by adding calls to `np.asarray`. I think @jakirkham had opened an issue on numpy stating that that particular function was slow. I think the conclusion was that it was simply slow due to requiring python introspection.

I'll try to get my previous PR through that talked about one usecase of dask for image processing
https://github.com/scikit-image/scikit-image/pull/3386

**Comment by GenevieveBuckley** [comment]
@jni was curious.

I personally suspect that there are not going to be very many places where it makes sense to worry about this, since it's often not a sensible idea to pass very large arrays directly into scikit-image functions & `apply_parallel()` exists anyway.

**Comment by JDWarner** [comment]
A lot of digital cameras exist today that are generating images up into the 50+ megapixel range, with 12 to 14 bit raw data.  I'm not sure how many of you use scikit-image on these sort of images, but the package performance is suboptimal there.  We do have `apply_parallel`, but Dask could also greatly assist.

In general, if we have the ability to be Dask compatible I think we should be.

**Comment by hmaarrfk** [comment]
@JDWarner I often find that dask array isn't the right tool for this kind of stuff. one is often better off using dask delayed on smaller chunks of data.

scikit-image is already compatible with dask delayed ;) which has worked pretty well for me :D

**Comment by hmaarrfk** [comment]
Also, if the intention is simply to make things "run", then we can discount performance issues that are brought up
https://github.com/scikit-image/scikit-image/pull/3823#discussion_r270497849

but ultimately, I think it is more beneficial to have a target user in mind. Is it the person using 256x256x3 images? or the one looking at 5kx5kx3 floats!

**Comment by hmaarrfk** [comment]
???? why was this closed?

**Comment by jni** [comment]
I agree, I think this should remain open so that we can revisit this discussion periodically. I think any Cython functions will not be portable, but anything that is bare-numpy might be made dask-compatible with minimal effort, especially in a future in which `__array_function__` becomes commonplace.

Also, this kind of work would benefit not just dask arrays, but also CuPy arrays.

**Comment by GenevieveBuckley** [comment]
> ???? why was this closed?

Sorry, I didn't realise people wanted to keep this open as a space for ongoing discussion, instead of being attached to a specific to-do. I apologize!

**Comment by GenevieveBuckley** [comment]
Here's a link to the `__array_function__` numpy enhancement proposal @jni mentions above: https://www.numpy.org/neps/nep-0018-array-function-protocol.html

**Comment by jakirkham** [comment]
> I think any Cython functions will not be portable

For these cases, it might be interesting to consider dispatching as discussed in issue ( https://github.com/scipy/scipy/issues/10204 ). Would be curious to hear thoughts on this idea. 🙂
